optaplanner-core/src/main/java/org/optaplanner/core/impl/score/inliner/ScoreInliner.java

```diff
@@ -16,18 +16,110 @@
     }
 
     public abstract Score_ extractScore(int initScore);
 
-    public abstract WeightedScoreImpacter buildWeightedScoreImpacter(Score_ constraintWeight);
+    public abstract WeightedScoreImpacter buildWeightedScoreImpacter(String constraintPackage, String constraintName,
+            Score_ constraintWeight);
+
+    protected Runnable addConstraintMatch(String constraintPackage, String constraintName, Score_ constraintWeight,
+            Score_ score, List<Object> justifications) {
+        String constraintId = ConstraintMatchTotal.composeConstraintId(constraintPackage, constraintName);
+        DefaultConstraintMatchTotal<Score_> constraintMatchTotal = constraintMatchTotalMap.computeIfAbsent(constraintId,
+                __ -> new DefaultConstraintMatchTotal<>(constraintPackage, constraintName, constraintWeight, zeroScore));
+        ConstraintMatch<Score_> constraintMatch =
+                constraintMatchTotal.addConstraintMatch(justifications, score);
+        List<DefaultIndictment<Score_>> indictmentList = justifications.stream()
+                .distinct() // One match might have the same justification twice
+                .map(justification -> {
+                    DefaultIndictment<Score_> indictment = indictmentMap.computeIfAbsent(justification,
+                            __ -> new DefaultIndictment<>(justification, zeroScore));
+                    indictment.addConstraintMatch(constraintMatch);
+                    return indictment;
+                }).collect(Collectors.toList());
+        return () -> {
+            constraintMatchTotal.removeConstraintMatch(constraintMatch);
+            if (constraintMatchTotal.getConstraintMatchSet().isEmpty()) {
+                constraintMatchTotalMap.remove(constraintId);
+            }
+            for (DefaultIndictment<Score_> indictment : indictmentList) {
+                indictment.removeConstraintMatch(constraintMatch);
+                if (indictment.getConstraintMatchSet().isEmpty()) {
+                    indictmentMap.remove(indictment.getJustification());
+                }
+            }
+        };
+    }
+
+    public Map<String, ConstraintMatchTotal<Score_>> getConstraintMatchTotalMap() {
+        // Unchecked assignment necessary as CMT and DefaultCMT incompatible in the Map generics.
+        return (Map) constraintMatchTotalMap;
+    }
+
+    public Map<Object, Indictment<Score_>> getIndictmentMap() {
+        // Unchecked assignment necessary as Indictment and DefaultIndictment incompatible in the Map generics.
+        return (Map) indictmentMap;
+    }
+
+    protected void ensureNonZeroConstraintWeight(Score_ constraintWeight) {
+        if (constraintWeight.equals(zeroScore)) {
+            throw new IllegalArgumentException("The constraintWeight (" + constraintWeight + ") cannot be zero,"
+                    + " this constraint should have been culled during node creation.");
```

Comment:
I presume this is an "impossible state" (AKA, the user should never see this no matter how much the user screws up - or put differently: it's definitely our fault), so let's have the error message start with "Impossible state: "
(see other exceptions in optaplanner codebase that start with "Impossible state:")